wrappers/services.go

```diff
@@ -705,41 +706,160 @@ func makeAbbrevWeekdays(start time.Weekday, end time.Weekday) []string {
-			switch week.Start.Pos {
-			case timeutil.EveryWeek:
+
+			if week.Start.Pos == timeutil.EveryWeek && week.End.Pos == timeutil.EveryWeek {
 				// eg: mon, mon-fri, fri-mon
 				days = append(days, fmt.Sprintf("%s *-*-*", abbrev))
-			case timeutil.LastWeek:
-				// eg: mon5
-				days = append(days, fmt.Sprintf("%s *-*~7/1", abbrev))
-			default:
-				// eg: mon1, fri1, mon1-tue2
-				startDay := (week.Start.Pos-1)*7 + 1
-				endDay := week.End.Pos * 7
-
-				// NOTE: schedule mon1-tue2 (all weekdays
-				// between the first Monday of the month, until
-				// the second Tuesday of the month) is not
-				// translatable to systemd.time(7) format, for
-				// this assume all weekdays and allow the runner
-				// to do the filtering
-				if week.Start != week.End {
+				continue
+			}
+			// examples:
+			// mon1 - Mon *-*-1..7 (Monday during the first 7 days)
+			// fri1 - Fri *-*-1..7 (Friday during the first 7 days)
+
+			// entries below are will make systemd timer expire more
+			// frequently than the schedule suggests, however snap
+			// runner evaluates current time and gates the actual
+			// action
+			//
+			// mon1-tue - *-*-1..7 *-*-8 (anchored at first
+			// Monday; Monday happens during the 7 days,
+			// Tuesday can possibly happen on the 8th day if
+			// the month started on Tuesday)
+			//
+			// mon-tue1 - *-*~1 *-*-1..7 (anchored at first
+			// Tuesday; matching Monday can happen on the
+			// last day of previous month if Tuesday is the
+			// 1st)
+			//
+			// mon5-tue - *-*~1..7 *-*-1 (anchored at last
+			// Monday, the matching Tuesday can still happen
+			// within the last 7 days, or on the 1st of the
+			// next month)
+			//
+			// fri4-mon - *-*-22-31 *-*-1..7 (anchored at 4th
+			// Friday, can span onto the next month, extreme case in
+			// February when 28th is Friday)
+			//
+			// XXX: since old versions of systemd, eg. 229 available
+			// in 16.04 does not support x..y ranges, days need to
+			// be enumerated like so:
+			// Mon *-*-1..7 -> Mon *-*-1,2,3,4,5,6,7
+			//
+			// XXX: old systemd versions do not suppor the last n
+			// days syntax eg, *-*~1, thus the range needs to be
+			// generated in more verbose way like so:
+			// Mon *-*~1..7 -> Mon *-*-21,22,23,24,25,26,27,28,29,30,31
+			// (21-28 is the last week, but the month can have
+			// anywhere from 28 to 31 days)
+			//
+			startPos := week.Start.Pos
+			endPos := startPos
+			if !week.AnchoredAtStart() {
+				startPos = week.End.Pos
+				endPos = startPos
+			}
+			startDay := (startPos-1)*7 + 1
+			endDay := (endPos) * 7
+
+			if week.IsSingleDay() {
+				// single day, can use the 'weekday' filter
+				if startPos == timeutil.LastWeek {
+					// last week of a month, which can be
+					// 21-28 in case of February, while
+					// month can have between 28 and 31 days
 					days = append(days,
-						fmt.Sprintf("*-*-%d..%d/1", startDay, endDay))
+						fmt.Sprintf("%s *-*-%s", abbrev, daysRange(22, 31)))
```

Comment:
Ok, so this can actually match 9 or 10 days for most months, but per your general comment above even if the schedule is active for longer period than it should, it's fine because we evaluate the schedule ourselves anyway right?

Reply:
Yes, the call pattern is `systemd -> snap.timer.*.timer -> snap.timer.*.service -> snap run --timer="<original-schedule>"`, where we evaluate `timeutil.Includes(schedule, time.Now())`